### PR TITLE
typescript migration

### PR DESCRIPTION
**Pull request by robostheimer**
Migrated ember-sinon-qunit over to typescript in order to make types available to consumers of this package.

All tests and linters pass.

ember ts:precompile passes

### CODE REVIEW

File: addon-test-support/index.ts [hunk 1]
```diff
@@ -1,5 +1,9 @@
 import { createSandbox, restoreSandbox } from './sinon-sandbox';
-
+declare global {
```
Comment (elwayman02) [inline addon-test-support/index.ts:2]: I think we might be able to replace this with a real import, actually:

`import * as QUnit from 'qunit';`

Comment (elwayman02) [reply addon-test-support/index.ts:2]: We would have to add `qunit` as a peerDependency of the package, like `sinon`. It probably already should have been one, to be honest.

Comment (elwayman02) [reply addon-test-support/index.ts:2]: If this works, we'd drop the `self.` from line 11, I think.

File: tests/helpers/assert-sinon-in-test-context.ts [hunk 2]
```diff
@@ -1,46 +1,66 @@
...
 
+interface TestContextExtended extends TestContext {
+  spy: SinonSpy;
+  stub: SinonStub;
+  mock: SinonSandbox['mock'];
+  fake: SinonFake;
+  replace: SinonSandbox['replace'];
+  replaceGetter: SinonSandbox['replaceGetter'];
+  replaceSetter: SinonSandbox['replaceSetter'];
+  sandbox: SinonSandBoxExtended;
```
Comment (elwayman02) [inline tests/helpers/assert-sinon-in-test-context.ts:25]: Where is `sandbox` used?

File: tests/unit/setup-sinon-test.ts [hunk 3]
```diff
@@ -13,12 +13,13 @@ module('Unit | ember-sinon-qunit | Setup in testStart/testDone', function () {
     let testDoneCalled = false;
 
     let qunit = {
-      testStart(callback) {
+      // eslint-disable-line
```
Comment (elwayman02) [inline tests/unit/setup-sinon-test.ts:16]: What is this disabling?

File: tsconfig.json [hunk 4]
```diff
@@ -0,0 +1,42 @@
+{
+  "extends": "@tsconfig/ember/tsconfig.json",
+  "compilerOptions": {
+    "experimentalDecorators": true,
+    "target": "es2017",
```
Comment (elwayman02) [inline tsconfig.json:5]: We can probably target something newer than `es2017`, as that's well out of our support matrix.

File: tsconfig.json [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    "allowJs": true,
+    "moduleResolution": "node",
+    "allowSyntheticDefaultImports": true,
+    "noImplicitAny": true,
+    "noImplicitThis": true,
+    "alwaysStrict": true,
+    "noFallthroughCasesInSwitch": true,
+    "noUnusedLocals": true,
+    "noImplicitReturns": true,
+    // change to true when all files are switched over to ts
```
Comment (elwayman02) [inline tsconfig.json:15]: Which files weren't converted?

File: package.json [hunk 6]
```diff
@@ -112,6 +112,7 @@
   },
   "peerDependencies": {
     "ember-source": "^3.28.0 || ^4.0.0",
+    "qunit": "^2.19.4",
```
Comment (elwayman02) [inline package.json:115]: Let's make this a bit more permissive and just do `^2.0.0` here.

Comment (elwayman02) [reply package.json:115]:
```suggestion
    "qunit": "^2.0.0",
```

File: tsconfig.json [hunk 7]
```diff
@@ -12,7 +12,6 @@
     "noFallthroughCasesInSwitch": true,
     "noUnusedLocals": true,
     "noImplicitReturns": true,
-    // change to true when all files are switched over to ts
```
Comment (elwayman02) [inline tsconfig.json]: Should we also change the value to true, per the comment being removed?

File: tests/helpers/assert-sinon-in-test-context.ts (hunk 2 above)
Comment (robostheimer) [reply tests/helpers/assert-sinon-in-test-context.ts:25]: line 124

File: tsconfig.json (hunk 5 above)
Comment (robostheimer) [reply tsconfig.json]: Comment was removed.  Left in by mistake.

File: package.json (hunk 6 above)
Comment (robostheimer) [reply package.json]: sure thing.

File: tsconfig.json (hunk 4 above)
Comment (robostheimer) [reply tsconfig.json]: es2020 work?

File: tests/unit/setup-sinon-test.ts (hunk 3 above)
Comment (robostheimer) [reply tests/unit/setup-sinon-test.ts]: vestige of development.  removed.

File: tsconfig.json (hunk 7 above)
Comment (robostheimer) [reply tsconfig.json]: Yes, made the change.  Compiling will be blocked by TS errors.

**Pull request merged**